### DISCUSSION TITLE
502 NO_STATUS_CODE_FROM_FUNCTION after deploying (10-second timeout issue?)

### DISCUSSION THREAD

**Discussion by lpatmo** [vercel/vercel#4103] [General] [answered]
Hi! I followed this tutorial to deploy an app: https://zeit.co/guides/deploying-next-and-mysql-with-now

I set up a Cloud SQL instance -- a 5.7 MySQL database.

The deployed app gives me the following error:

![image](https://user-images.githubusercontent.com/4512699/79711415-9b394a00-827c-11ea-9d35-ec67120da1a5.png)

If I run the app locally with `now dev`, all looks well:
![image](https://user-images.githubusercontent.com/4512699/79711460-b0ae7400-827c-11ea-97b1-b26f36fbe17c.png)

One thing the tutorial didn't mention was https://err.sh/now/missing-env-file, so I have an `.env` file that looks something like this:
```
MYSQL_HOST = 11.111.11.11.111
MYSQL_USER = exampleuser
MYSQL_PASSWORD = examplepassword
MYSQL_DATABASE = db
```

And my now.json file looks like this:

```
{
  "env": {
    "MYSQL_HOST": "@mysql_host",
    "MYSQL_USER": "@mysql_user",
    "MYSQL_PASSWORD": "@mysql_password",
    "MYSQL_DATABASE": "@mysql_database"
  }
}
```

I don't have environment variables set in my zeit.co project account. (I thought I needed to, but I think the .env file takes care of that?)

My zeit.co project profile also shows that my API requests all timed out after 10.01 seconds:
![image](https://user-images.githubusercontent.com/4512699/79711703-6a0d4980-827d-11ea-9d72-236f17c2a8bb.png)

My logs from production did not look very helpful: 
![image](https://user-images.githubusercontent.com/4512699/79711799-bbb5d400-827d-11ea-8ca4-794ff8cebba3.png)

Finally, here is the top of my `pages/index.js` file: 

![image](https://user-images.githubusercontent.com/4512699/79711839-eb64dc00-827d-11ea-8420-05fddb5bb065.png)

**Comment by bigsee** [comment]
I have the exact same symptom in production, except trying to POST to a MongoDB database within a serverless function.

The first request always times out at 10.01s and returns 502. The second request always works and returns 200 in a second or so.

![Screen Shot 2020-04-21 at 17 45 59](https://user-images.githubusercontent.com/11768967/79885970-3c004480-83f8-11ea-9b79-82f94c945016.png)

Did you find a solution?

**Reply by lpatmo** [reply]
Not yet! I was thinking of trying out AWS instead of Cloud SQL to make sure it's not just the Google service being slow, but haven't done that yet. :)

**Reply by bigsee** [reply]
Interesting... 🤔 Since you're using Cloud SQL / Google and I'm using mLab / Mongo DB, it seems possible that the issue is a configuration one at the Zeit/Vercel end.

I wonder if there is some way of configuring timeouts for serverless functions. I also found it interesting that I get the timeout on every other call, i.e. this pattern:

Call 1 - 502
Call 2 - 200
Call 3 - 502
Call 4 - 200

...and so on. Seems very odd to me...

**Comment by coetry** [comment]
A `.env` file only works for your local development environment. We recently launched the [Environment Variables UI](https://vercel.com/blog/environment-variables-ui) where you should place all your Production, Preview, and Development variables.

**Reply by lpatmo** [reply]
Thanks @coetry! I made sure to add the environment variables to production too in "settings", but alas it did not solve the issue I reported originally:

![image](https://user-images.githubusercontent.com/4512699/80048529-4dfbe900-84c5-11ea-9500-f63367b53fe3.png)

![image](https://user-images.githubusercontent.com/4512699/80048713-cbbff480-84c5-11ea-8a33-122af0d0b4a9.png)

**Reply by bigsee** [reply]
@lpatmo yes, agree with you. I'm don't think the original issue (10-second timeout) is about `.env` variables, although @coetry is of course right about needing to store them either using the UI or in the old way with `now secrets`.

@coetry any ideas on the original timeout issue with serverless functions?

**Comment by bigsee** [comment]
@lpatmo any luck finding a solution to the timeout issue?

Anyone from the Vercel team got any ideas at all on this? It's a real pain having a simple serverless API that doesn't respond consistently. The workarounds are adding a lot of unnecessary code for something that might have a simple cause...

**Comment by coetry** [comment, marked as answer]
@lpatmo @bigsee we have a timeout of 10s for functions on our Hobby plan and 60 seconds for functions on our Pro plan. If you need a larger timeout, then consider upgrading to a paid plan.

We have our limits documented [here](https://vercel.com/docs/v2/platform/limits).

A 10s timeout is the most we can sponsor for a free user :slightly_smiling_face: .

**Comment by helloitsm3** [comment]
I'm having the exact problem when trying to query to my MongoDB using Axios. It's impossible for my application to take more than 5 seconds to query 1 database. Do I really have to switch to other hosting providers?

**Reply by davidfurlong** [reply]
it seems to be the timeout on the serverless function rather than the database query.

**Reply by helloitsm3** [reply]
I've checked and it is due to my serverless function timing out. However, I can't seem to find a fix for this and ended up creating my own Express server to query from the database

**Comment by davidfurlong** [comment]
So I found this issue. Although I setup the google cloud integration for SQL the setup didn't tell me I had to whitelist the vercel IPs via cloud sql => connections => public ip, add => 0.0.0.0/0. See https://github.com/vercel/vercel/discussions/4595

This led to the timeouts
